### PR TITLE
[compiler-rt][ASan] Add function copying annotations

### PR DESCRIPTION
**Pull request by AdvenamTacet**
This PR adds a `__sanitizer_copy_contiguous_container_annotations` function, which copies annotations from one memory area to another. New area is annotated in the same way as the old region at the beginning (within limitations of ASan).

Overlapping case: The function supports overlapping containers, however no assumptions should be made outside of no false positives in new buffer area. (It doesn't modify old container annotations where it's not necessary, false negatives may happen in edge granules of the new container area.) I don't expect this function to be used with overlapping buffers, but it's designed to work with them and not result in incorrect ASan errors (false positives).

If buffers have granularity-aligned distance between them (`old_beg % granularity == new_beg % granularity`), copying algorithm works faster. If the distance is not granularity-aligned, annotations are copied byte after byte.

```cpp
void __sanitizer_copy_contiguous_container_annotations(
    const void *old_storage_beg_p, const void *old_storage_end_p,
    const void *new_storage_beg_p, const void *new_storage_end_p) {
```

This function aims to help with short string annotations and similar container annotations. Right now we change trait types of `std::basic_string` when compiling with ASan and this function purpose is reverting that change as soon as possible.

https://github.com/llvm/llvm-project/blob/87f3407856e61a73798af4e41b28bc33b5bf4ce6/libcxx/include/string#L738-L751

The goal is to not change `__trivially_relocatable` when compiling with ASan. If this function is accepted and upstreamed, the next step is creating a function like `__memcpy_with_asan` moving memory with ASan. And then using this function instead of `__builtin__memcpy` while moving trivially relocatable objects.

https://github.com/llvm/llvm-project/blob/11a6799740f824282650aa9ec249b55dcf1a8aae/libcxx/include/__memory/uninitialized_algorithms.h#L644-L646

---

I'm thinking if there is a good way to address fact that in a container the new buffer is usually bigger than the previous one. We may add two more arguments to the functions to address it (the beginning and the end of the whole buffer.

Another potential change is removing `new_storage_end_p` as it's redundant, because we require the same size.

Potential future work is creating a function `__asan_unsafe_memmove`, which will be basically memmove, but with turned off instrumentation (therefore it will allow copy data from poisoned area).

### CODE REVIEW

**Comment by philnik777** [comment]
trivial relocation is also interesting for erasing object from a vector, so the builtins should probably be more akin to a `memmove` than a `memcpy`.

**Comment by vitalybuka** [comment]
LGTM for Asan however HWASAN story is not clear.
Let's make sure we can use this for HWASAN as well. 

It's probably fine, but I'd like to think about this more.

**Comment by AdvenamTacet** [comment]
I still have to fix failing CI, I can sit to it on Monday and, if necessary, spend next week enough of time to fix it. I expect the problem to be a small implementation bug. Let me know if you have any high level requests for that code, I'm happy to implement them.
I will also create a separate PR cleaning compiler-rt ASan code, as I noticed a little bit of outdated code while writing this one.

> trivial relocation is also interesting for erasing object from a vector, so the builtins should probably be more akin to a memmove than a memcpy.

Should we change `__builtin_memcpy` to `__builtin_memmove`? Then I would have to add support for overlapping areas to this PR. (It complicates the implementation a little bit.)

> It's probably fine, but I'd like to think about this more.

Sure, let me know what are your conclusions!

**Comment by philnik777** [comment]
> > trivial relocation is also interesting for erasing object from a vector, so the builtins should probably be more akin to a memmove than a memcpy.
> 
> Should we change `__builtin_memcpy` to `__builtin_memmove`? Then I would have to add support for overlapping areas to this PR. (It complicates the implementation a little bit.)

Yes. My plan is to change the `__builtin_memcpy` to `__builtin_memmove` in `__uninitialized_allocator_relocate` to implement more optimizations in `vector`.

**Review by vitalybuka — CHANGES_REQUESTED** [review]
I assume work in progress.

**Comment by AdvenamTacet** [comment]
This PR is ready for the review. I see nothing more what has to be done in it.

Next step (but it may be in a separate PR) is adding a function like `asan_unsafe_memmove`, which would be a `memmove` with turned off instrumentation (to not trigger ASan error). I'm unsure in which file we should add the function and I'm open to suggestions. Optionally we can create `asan_unsafe_memmove_with_annotations` (to copy content and annotations in one call).

**Review by vitalybuka — COMMENTED** [review]
I need some time to reasonably review PR.

File: compiler-rt/lib/asan/asan_poisoning.cpp [hunk 1]
```diff
@@ -576,6 +576,235 @@ void __sanitizer_annotate_double_ended_contiguous_container(
...
+  if (!flags()->detect_container_overflow)
+    return;
+
+  VPrintf(2, "contiguous_container_src: %p %p\n", src_begin_p, src_end_p);
+  VPrintf(2, "contiguous_container_dst: %p\n", dst_begin_p);
+
+  uptr src_storage_begin = reinterpret_cast<uptr>(src_begin_p);
+  uptr src_storage_end = reinterpret_cast<uptr>(src_end_p);
+  uptr dst_storage_begin = reinterpret_cast<uptr>(dst_begin_p);
+  uptr dst_storage_end =
```
[the remark this answers is not in the document]
Comment (AdvenamTacet) [reply compiler-rt/lib/asan/asan_poisoning.cpp]: That was my first implementation and original reasoning, but later I changed it. My motivation behind removing that argument was intended use of that function when (without ASan) memmove or memcpy is used and those functions assume same size of both containers. But I'm happy with either, reverted that change for now.

Also, should we keep this order of arguments (src, dst) or should we change the order to dst be before src? I based order on old/new container annotation functions, like:
```
void __sanitizer_annotate_double_ended_contiguous_container(
    const void *storage_beg_p, const void *storage_end_p,
    const void *old_container_beg_p, const void *old_container_end_p,
    const void *new_container_beg_p, const void *new_container_end_p) {
```
But maybe it's better to mimic memmove/memcpy with destination being first?

File: compiler-rt/test/asan/TestCases/copy_container_annotations.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  return result;
+}
+
+void TestNonOverlappingContainers(size_t capacity, size_t off_old,
+                                  size_t off_new, int poison_buffers) {
+  size_t old_buffer_size = capacity + off_old + kGranularity * 2;
+  size_t new_buffer_size = capacity + off_new + kGranularity * 2;
+  char *old_buffer = new char[old_buffer_size];
```
[the remark this answers is not in the document]
Comment (AdvenamTacet) [reply compiler-rt/test/asan/TestCases/copy_container_annotations.cpp]: Does it look good now? Or should I change more?

**Review by AdvenamTacet — COMMENTED** [review]
> I need some time to reasonably review PR.

If I can help in any way, let me know!

**Review by vitalybuka — APPROVED** [review]
Please pull the rest of https://github.com/llvm/llvm-project/pull/110941 and LGTM

I guess future TODO is to optimize  SLOW routines, which is 15x slower then fast path.

I think reasonably readable and efficient approach is to create
iterators which walk src shadow in bytes and convert into bitsets of poisoned bytes, with required shifts by offset.

and another iterator which stores bitsets into dst shadow.

**Pull request merged**